### PR TITLE
Improve book, add diagrams

### LINKED ISSUE #327: Incorporate diagrams on my computer into the book
Just a little reminder for me, I got some useful diagrams which explain the ECS concept on my computer and I want to add them to the book.

### PR DESCRIPTION
**Pull request by torkleyy**
Fixes #327

[Rendered](https://torkleyy.github.io/specs/)

---
This change is [Reviewable](https://reviewable.io/reviews/slide-rs/specs/412)

### CODE REVIEW

**Comment by torkleyy** [comment]
Ah yeah, I stole your idea @WaDelma :smile:

**Comment by torkleyy** [comment]
When I made a presentation on ECS last year, I needed to explain the relationship and tables did work very good :)

File: book/src/01_intro.md [hunk 1]
```diff
@@ -60,31 +60,46 @@ public class Npc extends Character {
 Now you have stuff duplicated; you would have to write mostly identical code for
 your player and the NPC, even though e.g. they both share a transform.
 
-This is where ECS comes into play: Components are *associated* with entities;
-you just insert some component, whenever you like. Also, you only define the functionality
-once: One System, taking a `Force` and a `Mass` - it produces a `Velocity`.
+<img src="./images/entity-component.svg" alt="Entity-component relationship" width="20%" style="float:left;margin-right:15px" />
 
-In fact, an entity does not even own the components; it's just
+This is where ECS comes into play: Components are *associated* with entities; you just insert some component, whenever you like.
```
Comment (WaDelma) [inline book/src/01_intro.md]: s/you just insert some component,/you can just insert components/

File: book/src/01_intro.md [hunk 2]
```diff
@@ -60,31 +60,46 @@ public class Npc extends Character {
...
+One entity may or may not have a certain component. You can see an `Entity` as an ID into component tables, as illustrated in the
```
Comment (WaDelma) [inline book/src/01_intro.md]: s/component/component associated with it/

File: book/src/01_intro.md [hunk 3]
```diff
@@ -60,31 +60,46 @@ public class Npc extends Character {
...
+diagram below. We could theoretically store all the components together with the entity, but that would be very inefficient;
```
Comment (WaDelma) [inline book/src/01_intro.md]: s/below/above/

File: book/src/01_intro.md (hunk 2 above)
Comment (WaDelma) [inline book/src/01_intro.md]: Should there be clarification that metaphors tables are referring to database tables? So that a reader unfamiliar with databases wont get confused.

File: book/src/01_intro.md [hunk 4]
```diff
@@ -60,31 +60,46 @@ public class Npc extends Character {
...
+[chapter 5]: ./05_storages.html
+
+<img src="./images/component-tables.svg" alt="Component tables" width="100%"  />
+
+`Entity 1` has each of those components, `Entity 2` only a `Force`, etc.
+
+Now we're only missing the last character in ECS - the "S" for `System`. Whereas components and entities are purely data,
+systems contain all the logic of your application. A system typically iterates over all entities that fulfill specific constraints,
+like "has both a force and a mass". Based on this data a system will execute code, e.g. produce a velocity out of the force and the mass.
+This is the additional advantage I wanted to point out with the `Player` / `Npc` example; in an ECS, you can simply add attributes and
```
Comment (WaDelma) [inline book/src/01_intro.md]: s/is/has/
s/simply add attributes/easily add new attributes to existing entities/

File: book/src/01_intro.md [hunk 5]
```diff
@@ -60,31 +60,46 @@ public class Npc extends Character {
...
+that's also how you define your behaviour (this is called [data-driven] programming).
```
Comment (WaDelma) [inline book/src/01_intro.md]: s/ and that's also how you define your/. Because the systems processing depends on the components associated with an entity this is also the way you define your entities/

File: book/src/01_intro.md [hunk 6]
```diff
@@ -60,31 +60,46 @@ public class Npc extends Character {
...
-which include both `A` and `B`.
+By simply adding a force to an entity that has a mass, you can make it move, because a `Velocity` will be produced for it.
 
 ## Where to use an ECS?
 
-In case you were looking for a general-purpose library for doing it
-the data-oriented way, I have to disappoint you; there is none.
-ECS libraries are best-suited for creating games; in fact, I've never
-heard of anybody using an ECS for something else.
+In case you were looking for a general-purpose library for doing it the data-oriented way, I have to disappoint you; there is none.
```
Comment (WaDelma) [inline book/src/01_intro.md]: s/is/are/

File: book/src/06_system_data.md [hunk 7]
```diff
@@ -102,3 +102,29 @@ There are actually three ways of doing this:
   that the order here is important, using `Read<'a, Option<Resource>>` will not 
   result in the same behavior (it will try to fetch `Option<Resource>` from `World`, 
   instead of doing an optional check if `Resource` exists).
+
+## Specifying `SystemData`
+
+As mentioned earlier, `SystemData` is implemented for tuples of up to 26 elements. Should you ever need
```
Comment (WaDelma) [inline book/src/06_system_data.md]: s/of //

File: book/src/06_system_data.md [hunk 8]
```diff
@@ -102,3 +102,29 @@ There are actually three ways of doing this:
...
+more, you could even nest these tuples. However, at some point you can't keep track of all the elements anymore.
```
Comment (WaDelma) [inline book/src/06_system_data.md]: s/you can't/it becomes hard to/
s/ anymore//

File: book/src/06_system_data.md [hunk 9]
```diff
@@ -102,3 +102,29 @@ There are actually three ways of doing this:
...
+That's why you can also create your own `SystemData` bundle using a struct:
```
Comment (WaDelma) [inline book/src/06_system_data.md]: Btw what happens if you try to derive it for enum? I realised you could try to do this after reading this.

File: book/src/01_intro.md (hunk 3 above)
Comment (torkleyy) [reply book/src/01_intro.md]: the one I'm referring to is below actually

File: book/src/06_system_data.md (hunk 9 above)
Comment (torkleyy) [reply book/src/06_system_data.md]: Doesn't work, which variant should it select?

Comment (WaDelma) [reply book/src/06_system_data.md]: Guessed as much :D

File: book/src/01_intro.md (hunk 3 above)
Comment (WaDelma) [reply book/src/01_intro.md]: Well then you are not referring to this one at all, right?

File: book/src/01_intro.md (hunk 4 above)
Comment (torkleyy) [reply book/src/01_intro.md]: Hmm these changes you suggest here, I don't really get how they improve the text. I think "is" fits better here because I want to regard specifically the advantage, not the way systems work.

As for the second change you propose, I chose "simply" to describe how plain the whole concept it is, yet it allows for such patterns. The difficulty of adding the attribute is not what I wanted to describe.

File: book/src/01_intro.md (hunk 2 above)
Comment (torkleyy) [reply book/src/01_intro.md]: That's what I had initially, but then the sentence sounded a bit unnatural and I didn't want to repeat "associated" as it's already used in the line before.

Comment (Xaeroxe) [reply book/src/01_intro.md]: I'm with @torkleyy here, the extra words don't add to comprehension.

File: book/src/01_intro.md (hunk 4 above)
Comment (Xaeroxe) [reply book/src/01_intro.md]: Personally I'd leave the "is"

I'd actually prefer a hybrid for the second request though

"simply add new attributes to entities"

File: book/src/01_intro.md (hunk 5 above)
Comment (Xaeroxe) [reply book/src/01_intro.md]: Suggestion seems a bit wordy and doesn't add any new information imo.

File: book/src/01_intro.md [hunk 10]
```diff
@@ -60,31 +60,46 @@ public class Npc extends Character {
...
+
+<img src="./images/component-tables.svg" alt="Component tables" width="100%"  />
+
+`Entity 1` has each of those components, `Entity 2` only a `Force`, etc.
+
+Now we're only missing the last character in ECS - the "S" for `System`. Whereas components and entities are purely data,
+systems contain all the logic of your application. A system typically iterates over all entities that fulfill specific constraints,
+like "has both a force and a mass". Based on this data a system will execute code, e.g. produce a velocity out of the force and the mass.
+This is the additional advantage I wanted to point out with the `Player` / `Npc` example; in an ECS, you can simply add new attributes
+to entities and that's also how you define your behaviour (this is called [data-driven] programming).
```
Comment (WaDelma) [inline book/src/01_intro.md]: I was changing the your here, because for me it sounds like you are talking about coders behaviour, but it's fine.

**Comment by torkleyy** [comment]
Ah missed that, thanks!

bors r=Xaeroxe,WaDelma
